python-api/livy/client.py

```diff
@@ -0,0 +1,444 @@
+        >>> def add_file_job(context):
+        >>>    from pyspark import SparkFiles
+        >>>    def func(iterator):
+        >>>        with open(SparkFiles.get("test_add.txt")) as testFile:
+        >>>        fileVal = int(testFile.readline())
+        >>>        return [x * fileVal for x in iterator]
+        >>>    return context.sc.parallelize([1, 2, 3, 4])
+        >>>             .mapPartitions(func).collect()
+
+        >>> client.submit(add_file_job)
+        """
+        return self._add_file_or_pyfile_job("add-file", file_uri)
+
+    def add_jar(self, file_uri):
+        """
+        Adds a jar file to the running remote context.
+
+        Note that the URL should be reachable by the Spark driver process. If
+        running the driver  in cluster mode, it may reside on a different host,
+         meaning "file:" URLs have to exist on that node (and not on the
+         client machine).
+
+        Parameters
+        ----------
+        file_uri : string, pathlib.PurePath or pathlib2.PurePath
+            Representation of path to a local file using filesystem encoding.
+
+        Returns
+        -------
+        future : concurrent.futures.Future
+            A future to monitor the status of the job
+
+        Examples
+        -------
+        >>> import pathlib2
+        >>> file_uri = pathlib2.Path("/test_package.jar").as_uri()
+
+        >>> client.add_jar(file_uri)
+
+        """
+        return self._add_file_or_pyfile_job("add-jar", file_uri)
+
+    def add_pyfile(self, file_uri):
+        """
+        Adds a .py or .zip to the running remote context.
+
+        Note that the URL should be reachable by the Spark driver process. If
+        running the driver  in cluster mode, it may reside on a different host,
+         meaning "file:" URLs have to exist on that node (and not on the
+         client machine).
+
+        Parameters
+        ----------
+        file_uri : string, pathlib.PurePath or pathlib2.PurePath
+            Representation of path to a local file using filesystem encoding.
+
+        Returns
+        -------
+        future : concurrent.futures.Future
+            A future to monitor the status of the job
+
+        Examples
+        -------
+        >>> import pathlib2
+        >>> file_uri = pathlib2.Path("/test_package.egg").as_uri()
+
+        >>> client.add_pyfile(file_uri)
+
+        >>> def add_pyfile_job(context):
+        >>>    # Importing module from test_package.egg
+        >>>    from test.pyfile_test import TestClass
+        >>>    test_class = TestClass()
+        >>>    return test_class.say_hello()
+
+        >>> client.submit(add_pyfile_job)
+        """
+        return self._add_file_or_pyfile_job("add-pyfile", file_uri)
+
+    def upload_file(self, open_file):
+        """
+        Upload a file to be passed to the Spark application.
+
+        Parameters
+        ----------
+        open_file : file
+            A handle to the local file to be uploaded.
```

Comment:
Is there any advantage in requiring an open file here? If so, you should clarify who's responsible for closing the file.

It seems the API would be easier to use if you just took a path (= string) instead.